Applications/Spire/Source/KeyBindings/KeyBindings.cpp

```diff
@@ -0,0 +1,85 @@
+#include "Spire/KeyBindings/KeyBindings.hpp"
+#include <algorithm>
+#include <type_traits>
+
+using namespace Nexus;
+using namespace Spire;
+
+void KeyBindings::set_binding(QKeySequence sequence, const Action& action) {
+  auto region = std::visit([] (auto& action) {
+    using Type = std::decay_t<decltype(action)>;
+    if constexpr(std::is_same_v<Type, CancelAction>) {
+      return Region(Region::GlobalTag());
+    } else {
+      return action.m_region;
+    }
+  }, action);
+  auto i = m_bindings.find(sequence);
+  if(i == m_bindings.end()) {
+    i = m_bindings.insert(sequence, Actions(boost::none));
+  }
+  i->Set(region, action);
+}
+
+void KeyBindings::reset_binding(const Region& region,
+    const QKeySequence& sequence) {
+  auto i = m_bindings.find(sequence);
+  if(i != m_bindings.end()) {
+    auto& actions = *i;
+    if(region.IsGlobal()) {
+      actions.Set(region, boost::none);
+    } else {
+      actions.Erase(region);
+      if(actions.GetSize() == 1 && actions.Get(Region(Region::GlobalTag()))) {
+        m_bindings.erase(i);
+      }
+    }
+  }
+}
+
+boost::optional<const KeyBindings::Action&>
+    KeyBindings::find_binding(const Region& region,
+    const QKeySequence& sequence) const {
+  auto i = m_bindings.find(sequence);
+  if(i == m_bindings.end()) {
+    return boost::none;
+  }
+  auto& action = i->Get(region);
+  if(action) {
```

Comment:
It might be a good idea to return `boost::optional<Action>` by copy instead. Optional references are not really safer than pointers, nor are they easier to comprehend and deal with. Furthermore, `std::optional` does not support references and there is a chance `boost::optional` will not as well later down the road.